src/Features/LanguageServer/Protocol/Extensions/Extensions.cs

```diff
@@ -23,31 +25,68 @@ public static Uri GetURI(this Document document)
             return ProtocolConversions.GetUriFromFilePath(document.FilePath);
         }
 
-        public static Document? GetDocumentFromURI(this Solution solution, Uri fileName, string? clientName = null)
+        public static ImmutableArray<Document> GetDocuments(this Solution solution, Uri uri, string? clientName = null)
         {
             // TODO: we need to normalize this. but for now, we check both absolute and local path
             //       right now, based on who calls this, solution might has "/" or "\\" as directory
             //       separator
-            var documentId = solution.GetDocumentIdsWithFilePath(fileName.AbsolutePath).FirstOrDefault() ??
-                             solution.GetDocumentIdsWithFilePath(fileName.LocalPath).FirstOrDefault();
+            var documentIds = solution.GetDocumentIdsWithFilePath(uri.AbsolutePath);
 
-            var document = solution.GetDocument(documentId);
+            if (!documentIds.Any())
+            {
+                documentIds = solution.GetDocumentIdsWithFilePath(uri.LocalPath);
+            }
+
+            var documents = documentIds.SelectAsArray(id => solution.GetRequiredDocument(id));
 
-            if (clientName != null)
+            // If we don't have a client name, then we're done filtering
+            if (clientName == null)
             {
-                var documentPropertiesService = document?.Services.GetService<DocumentPropertiesService>();
+                return documents;
+            }
+
+            // We have a client name, so we need to filter to only documents that match that name
+            return documents.WhereAsArray(document =>
+            {
+                var documentPropertiesService = document.Services.GetService<DocumentPropertiesService>();
+
                 // When a client name is specified, only return documents that have a matching client name.
                 // Allows the razor lsp server to return results only for razor documents.
                 // This workaround should be removed when https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1106064/
                 // is fixed (so that the razor language server is only asked about razor buffers).
-                if (!Equals(documentPropertiesService?.DiagnosticsLspClientName, clientName))
+                return Equals(documentPropertiesService?.DiagnosticsLspClientName, clientName);
+            });
+        }
+
+        public static Document? GetDocument(this Solution solution, TextDocumentIdentifier documentIdentifier, string? clientName = null)
+        {
+            var documents = solution.GetDocuments(documentIdentifier.Uri, clientName);
+
+            if (documents.Length == 0)
+            {
+                return null;
+            }
+
+            if (documents.Length > 1)
```

Comment:
I guess that's another good question for them. I was being cautious as you can see, but there's potentially a race: if the contexts change on the server and the client queries using an old context, those requests would then fail until the client knows new contexts again. If there was a way to communicate that failure so they knew they should refresh that would be good, but I'm not sure we have a way to communicate that today.

Reply:
Would throwing an exception be a way for the client to know they need to refresh?  Either way, I think I would prefer throwing - that would at least indicate that something is wrong rather than just having the client continue to chug along with an incorrect context.  I feel like if we just continue we may run into hard to diagnose issues with random project contexts being used.

Reply:
Let's continue with this approach for safety for now but I'll raise that question too on the mail thread.